### PR TITLE
add `boa.env.time_travel()`

### PR DESCRIPTION
**Pull request by bout3fiddy**
easy option to mine a single block (with a user specified block timestamp, set to default at 12 sec: or 11?)

also added block_timestamp and block_number properties.

### CODE REVIEW

File: boa/environment.py [hunk 1]
```diff
@@ -470,9 +478,9 @@ class FakeMessage(Message):  # Message object with settable attrs
         return self.vm.state.computation_class.apply_message(self.vm.state, msg, tx_ctx)
 
     # function to time travel
-    def time_travel(self, sleep_time):
-        self.vm.patch.timestamp += sleep_time
-        self.vm.patch.block_number += sleep_time // 12
+    def time_travel(self, time_delta: int, block_delta: int = 12):
```
Comment (charles-cooper) [inline boa/environment.py:481]: i think the implementation should be more like
```suggestion
    def time_travel(self, seconds = None, blocks = None):
        if seconds is None == blocks is None:
            raise ValueError("one of seconds or blocks should be set")
        if seconds is not None:
            blocks = seconds // 12
        else:
            seconds = blocks * 12
        self.vm.patch.timestamp += seconds
        self.vm.patch.block_number += blocks
```

File: boa/environment.py [hunk 2]
```diff
@@ -266,6 +266,14 @@ def __init__(self):
...
         self._contracts = {}
 
+    @property
+    def timestamp(self):
+        return self.vm.state.timestamp
+
+    @property
+    def block_number(self):
+        return self.vm.state.block_number
+
```
Comment (charles-cooper) [inline boa/environment.py:276]: not sure these should have "special" treatment relative to other vm patchables. better to do a more full rehaul if/when we address https://github.com/vyperlang/titanoboa/issues/32

**Review by charles-cooper — CHANGES_REQUESTED** [review]

File: boa/environment.py [hunk 3]
```diff
@@ -470,9 +470,20 @@ class FakeMessage(Message):  # Message object with settable attrs
...
     # function to time travel
-    def time_travel(self, sleep_time):
-        self.vm.patch.timestamp += sleep_time
-        self.vm.patch.block_number += sleep_time // 12
+    def time_travel(self, seconds=None, blocks=None, block_delta=12):
+
+        if seconds is None and blocks is None:
+            raise ValueError("One of seconds or blocks should be set")
+        if seconds is not None:
+            if seconds <= 0:
```
Comment (charles-cooper) [inline boa/environment.py:478]: i think time traveling to the past should be allowed

**Review by charles-cooper — CHANGES_REQUESTED** [review]

File: boa/environment.py (hunk 3 above)
Comment (bout3fiddy) [reply boa/environment.py:478]: Hmm can you give me an example where this would be ok?

File: boa/environment.py [hunk 4]
```diff
@@ -470,9 +470,22 @@ class FakeMessage(Message):  # Message object with settable attrs
...
-        self.vm.patch.timestamp += sleep_time
-        self.vm.patch.block_number += sleep_time // 12
+    def time_travel(
+        self,
+        seconds: Optional[int] = None,
+        blocks: Optional[int] = None,
+        block_delta: int = 12,
+    ):
+
+        if not (seconds or blocks):
```
Comment (charles-cooper) [inline boa/environment.py:480]: i think 0 value for seconds or blocks is ok, just like negative numbers. so i think a clean check here is `if seconds is None == blocks is None`

File: boa/environment.py [hunk 5]
```diff
@@ -470,9 +470,22 @@ class FakeMessage(Message):  # Message object with settable attrs
...
+            raise ValueError("One of seconds or blocks should be set")
+        if seconds and not blocks:
+            blocks = seconds // block_delta
+        elif not seconds and blocks:
+            seconds = blocks * block_delta
```
Comment (charles-cooper) [inline boa/environment.py:485]: with the guard for `seconds is not None != blocks is not None`, these conditions can be simplified - to `if seconds is not None`, and `else`.

Comment (charles-cooper) [reply boa/environment.py:485]:
```suggestion
        if seconds is not None:
            blocks = seconds // block_delta
        else:
            seconds = blocks * block_delta
```

File: boa/environment.py (hunk 4 above)
Comment (charles-cooper) [reply boa/environment.py:480]:
```suggestion
        if seconds is None == blocks is None:
```

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**